### ISSUE TITLE
Rework `endpoint.state` to use a dedicated type

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#14841]
The `state` field of the `Endpoint` object directly uses the `string` type, but it is expected to only hold specific values (like an enum). With it not being strongly typed, it's possible to call `Endpoint.SetState("foo")` and arbitrarily configure the states, rather than using one of a well-known set of states.

This cleanup request is to add a new type for the state and force that callers make use of the dedicated type when interacting with functions like `GetState()` and `SetState()`, to reduce the likelihood that it gets passed random nonsense values.

**Comment by ammmk** [comment]
@joestringer, I'd like to work on this, could you assign it to me?

**Comment by pchaigno** [comment]
@ammmk Done. Don't hesitate to reach out on Slack (e.g., #development) if you have any question.

**Comment by ammmk** [comment]
I just wrapped all the states in the new string-based type.
That doesn't prevent from putting string constants and literals `Endpoint.SetState("foo")` into, but helps with variables and typed constants.

I can go further and change the State to a numeric based type, which requires introducing additional mapping with `models.EndpointState`.
That will handle `Endpoint.SetState("foo")`, but still wouldn't prevent from putting invalid literal.

Let me know if I've got it right.
